### ISSUE TITLE
Implement script for extracting features

### ISSUE THREAD

**Issue by AlekseySh** [OML-Team/open-metric-learning#85]

**Comment by churnikov** [comment]
Hi!

I'd like to work on this issue. I hope it's not done yet :)

**Comment by AlekseySh** [comment]
@churnikov you are welcome! @DaloroAT will provide you with a road map today or tomorrow

**Comment by churnikov** [comment]
Sounds good :)

I've looked through contributions guide and I hope you'll give me details on what you expect :)

**Comment by DaloroAT** [comment]
Hey @churnikov !

We can split this script into several parts:
1) Read config, extract parameters about the model, transforms, dataloader and folder with images
2) Prepare `ListDataset` for process list of images
3) Substitute this dataset and iterate with dataloader
4) Save features into some file

We have no `ListDataset` in our repo now so we can start from point 2. After implementing and merging the code with `ListDataset` we can return to the rest of the points and discuss the design.

---

## Design

`ListDataset` should be very simple. It can iterate over files, apply transforms, and return tensors.

``` python
from torch.utils.data import Dataset
import torch

class ListDataset(Dataset):
    def __init__(self, filenames_list, transforms, f_imread):
        ...

    def __getitem__(self, idx) -> torch.Tensor:
        ...
```

Please, don't forget to support different types of transforms. Now we use transforms from `albumentations` and default `torchvision` transforms, you can check it in `oml.datasets.base.BaseDataset`.

Transforms are responsible for augmentations of the original images, resizing and normalizing.

## Test

Iterate over the files in mock dataset with dataloader and check shapes of batches. 

Mock dataset can be found in 
``` python
from oml.const import MOCK_DATASET_PATH
print(MOCK_DATASET_PATH / 'images')
```
If you have no mock dataset locally, use `make download_mock_dataset` from the root of the project.

---

What do you think @AlekseySh ?

**Comment by AlekseySh** [comment]
@DaloroAT Agree, let's start by implementing `ListDataset` in a dedicated PR. Then we can finish the rest of the functionality. 

 if the task is clear for you, @churnikov ?

**Comment by churnikov** [comment]
Sounds good :)

**Comment by DaloroAT** [comment]
Great!

I created a separate issue #205 for this particular task. Let's continue there @churnikov @AlekseySh .

We'll come back here when we're done with the dataset.

**Comment by churnikov** [comment]
Hi!

I think, as #205 is done, we can continue with this task 😄

**Comment by DaloroAT** [comment]
Great! Thank you @churnikov 

Now we have all components to implement that script. 

Let's assume that our script's goal is to extract images' features and save them on some file. 
Sources of images are the following:
1) Specific folder.
2) CSV with paths and possible bounding boxes.

Place your solution into the `examples` folder at the root of the project.

I'm not sure about the format of the features file, but  `json` is supported across different languages and has a good standard. It's not optimal to store arrays, but now we have no real cases for that script. So we can add support for other formats in the future.

## Config

We need to create and support a config with the parameters:

``` yaml
images_folder: ...
dataframe_name: ...
features_file: ...

batch_size: ...
num_workers: ...

transforms:
    ...

model: 
    ...
```

## Features file

The structure of `json` is the following:

```  yaml
{
"images_folder": <from config>,
"dataframe_name": <from config>,
"model": <nested dict from config section>,
"transforms": <nested dict from config section>,
"filenames": [file1, fil2, ...],
"features": [vector1, vector2, ...]
```

- `fileK` - absolute path to the Kth image.
- `vectorK` - list with float values [0.123, 1.356, ...] for Kth image.

## Test

Scenario:
1) Write 2 configs.
2) Inference on the folder with the mock dataset.
3) Inference with using CSV to mock dataset.
4) Compare pairs of `(file, vector)` between approaches and make sure that features are equal.

Check [example1](https://github.com/OML-Team/open-metric-learning/blob/main/tests/test_examples/test_runs_via_configs/val_mock.py) and [example2](https://github.com/OML-Team/open-metric-learning/blob/main/tests/test_examples/test_runs_via_configs/test_via_configs.py) how to run tests with configs.

## Check list

1) Parameters `images_folder` and `dataframe_name` are not specified simultaneously. One should be `null` if both are presented or only one should be in config.
2) If CSV, check that all files exist.
3) If a folder, use `rglob` to obtain nested files.
4) Check, that all files can be read. You can check it without decoding with [PIL](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.verify). It's not 100% guaranteed that a file can be read, but better than nothing.
5) If CSV, validate names of available columns (you can check it in `const.py`) and suggest to the user the correct names.
6) You can ignore PyTorch Lightning and implement it on vanilla PyTorch.
7) Add docs, we can discuss details after implementing core functionality.
---

Would you like to add something @AlekseySh ?

**Comment by AlekseySh** [comment]
@DaloroAT I am good.
But it's not true:
>  now we have no real cases for that script.
basically, I implemented a very dirty draft of it by myself and used it in my last experiments :)

**Comment by DaloroAT** [comment]
Oh, new PR)

Anyway this part of my comment was devoted to format of file for users, who need to extract and save for some purposes

**Comment by DaloroAT** [comment]
But if you found use case, we can add that to task

**Comment by AlekseySh** [comment]
I use pickle for myself, but it's not language-agnostic.
We can implement it with JSON, and then check if it can handle features for the InShop dataset (for example). 
If it will be inconvenient, we can think about hdf5 files..

**Comment by churnikov** [comment]
Hi @DaloroAT 
I'm almost done and approaching documentation
https://github.com/OML-Team/open-metric-learning/pull/234#issuecomment-1339021578

You suggested to discuss docs at that point.

**Comment by DaloroAT** [comment]
Hey @churnikov 

Let's discuss docs. 

We will add a new section in the big examples section, you can check the structure by [link](https://open-metric-learning.readthedocs.io/en/latest/examples/config.html#). You can create separate markdown snippet, and place them [there](https://github.com/OML-Team/open-metric-learning/tree/main/docs/source/examples). 

In general, you should highlight the following points:
- Which source of data user can use for inference, and how to prepare data for [some cases](https://open-metric-learning.readthedocs.io/en/latest/examples/config.html#usage-with-custom-dataset)?
- Parameters (transforms) for inference. Probably give the link to some default transform with normalizing and resizing with keeping the aspect ratio.
- Describe the structure of the output file.

Start with how you would like to see the manual as a lazy external user, then we can add some details during review.
